### PR TITLE
Rename the doc target.

### LINKED ISSUE #689: Make documentation building optional, as it can clash when the project is used as submodule
# Summary
At the moment, Documentation is built by default if Doxygen is found and the target is named the generic `doc`: https://github.com/oneapi-src/oneDNN/blob/master/cmake/Doxygen.cmake#L63

This causes issues such as:
`CMake Error at CMakeLists.txt:437 (add_custom_target):
  add_custom_target cannot create target "doc" because another target with
  the same name already exists.  The existing target is a custom target
  created in source directory`
When using OneDNN as a submodule in another project.

Instead, there should be an option such as DNNL_BUILD_DOC, or change the target to `dnnl_doc`, so that it doesn't clash with another generic custom target name.

### PR DESCRIPTION
**Pull request by XapaJIaMnu**
Fixes #689.

This is the simplest solution that preserves the functionailty and solves the problem.

### CODE REVIEW

File: cmake/Doxygen.cmake [hunk 1]
```diff
@@ -60,7 +60,7 @@ if(DOXYGEN_FOUND)
         COMMAND ${CMAKE_COMMAND} -E touch ${DOXYGEN_STAMP_FILE}
         WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
         COMMENT "Generating API documentation with Doxygen" VERBATIM)
-    add_custom_target(doc DEPENDS ${DOXYGEN_STAMP_FILE})
+    add_custom_target(oneDNNdoc DEPENDS ${DOXYGEN_STAMP_FILE})
```
Comment (rsdubtso) [inline cmake/Doxygen.cmake:63]: Sorry, I should've been more clear in the issue text. The acceptable solution is to have a cmake option. Changing the target name will break lots of things (our automation, packagers' scripts, etc).

Comment (emfomenk) [reply cmake/Doxygen.cmake:63]: @rsdubtso, what do you think if we enable this target only if oneDNN is not a sub-project of another project?

Comment (rsdubtso) [reply cmake/Doxygen.cmake:63]: I would be totally fine with that (we could use the same logic for tests, etc). I just could not find a reliable way to do this (but this does not mean that there is no way to do this).

**Comment by emfomenk** [comment]
@XapaJIaMnu,

Could you please validate the following fix?

``` patch
diff --git a/cmake/Doxygen.cmake b/cmake/Doxygen.cmake
index 2c1631813..0b7a09c8b 100644
--- a/cmake/Doxygen.cmake
+++ b/cmake/Doxygen.cmake
@@ -60,7 +60,9 @@ if(DOXYGEN_FOUND)
         COMMAND ${CMAKE_COMMAND} -E touch ${DOXYGEN_STAMP_FILE}
         WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
         COMMENT "Generating API documentation with Doxygen" VERBATIM)
-    add_custom_target(doc DEPENDS ${DOXYGEN_STAMP_FILE})
+    if(DNNL_IS_MAIN_PROJECT)
+        add_custom_target(doc DEPENDS ${DOXYGEN_STAMP_FILE})
+    endif()

     if(NOT DNNL_INSTALL_MODE STREQUAL "BUNDLE")
         install(
diff --git a/cmake/utils.cmake b/cmake/utils.cmake
index d9efd1854..564276302 100644
--- a/cmake/utils.cmake
+++ b/cmake/utils.cmake
@@ -23,6 +23,10 @@ endif()
 set(utils_cmake_included true)
 include("cmake/options.cmake")

+if ("${CMAKE_PROJECT_NAME}" STREQUAL "${PROJECT_NAME}")
+    set(DNNL_IS_MAIN_PROJECT TRUE)
+endif()
+
 # Common configuration for tests / test cases on Windows
 function(maybe_configure_windows_test name kind)
     if(WIN32 AND (NOT DNNL_BUILD_FOR_CI))
```

**Comment by XapaJIaMnu** [comment]
@emfomenk yes I can confirm this works. This should probably be extended to the tests and other optionally compiled components.

**Comment by emfomenk** [comment]
@XapaJIaMnu, I am fine with removing the tests and examples (that populate lots of _optional_ targets). However, I am reluctant to remove the `install` target that might be used by someone already. Would it be fine for you if we keep it as is?

**Comment by emfomenk** [comment]
And thanks for the validation, btw :)

**Comment by emfomenk** [comment]
After internal discussion we decided that we don't want to change the current behavior wrt to the tests. So the recommendation here is to include oneDNN project with pre-defined options `DNNL_BUILD_EXAMPLES=OFF` and `DNNL_BUILD_TESTS=OFF`.

With that I am going to promote the patch as it is shown above.

**Comment by XapaJIaMnu** [comment]
@emfomenk , thank you for your reply and quick solution to the problem. I am ok with building examples and tests by default, since I can use the flags to disable them. The only thing that I couldn't disable without out-of-tree patches were the docs.

Cheers

**Comment by emfomenk** [comment]
@XapaJIaMnu, the change is promoted. Thanks for the report and help with the fix!

**Pull request closed without merging**